src/osd/ReplicatedPG.h

```diff
@@ -493,6 +493,7 @@ class ReplicatedPG : public PG, public PGBackend::Listener {
     bool cache_evict;     ///< true if this is a cache eviction
     bool ignore_cache;    ///< true if IGNORE_CACHE flag is set
     bool ignore_log_op_stats;  // don't log op stats
+    bool unmodified_omap;  // false if modification of omap data
```

Comment:
This is a missed opportunity for abstraction.  unmodified_omap/unmodified_extents in OpContext here have precisely the same rules as the corresponding entries in both the pg log entries and the pg missing set items.  Please create a class encapsulating these two elements with appropriate methods like:
```
class ObjectCleanRegions {
  bounded_lossy_interval_set<uint64_t> clean_offsets;
  bool clean_omap;
public:
  void merge(const ObjectCleanRegions &other);
  void mark_data_region_dirty(uint64_t offset, uint64_t len);
  interval_set<uint64_t> get_dirty_regions() const;
  bool omap_is_dirty() const;
};
```

This would allow you to unit test this logic (particularly the merge logic) and avoid duplicating it everywhere you need to combine these entries.  It also allows ReplicatedPG::do_op to be clearer about what it is doing and makes it less likely that later modifications will misuse these structures.  This structure would also take care (either by using a bounded_lossy_interval_set as I did above or by using an interval_set directly) of limiting its own in memory size.